Refacto combination suppliers

jolelievre opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Refacto of the combination suppliers:<br>- suppliers selection is removed from the combination form, it will always be in the product form now in Options tab<br>- the edition of product suppliers details though remain in each combination form to avoid having thousands of rows to edit in product form<br>- the js suppliers manager is split into two different and independent components which provide public setters and/or callback to handle sync with other components<br>- thus the ProductFormModel is no longer needed to be injected in the components which was a bad pattern from the start
| Type?             | refacto
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Fixes #27034 and #26906 and #23002
| How to test?      | Test the suppliers for a standard product it should work as usual, for a combination you now edit the associations in the Options tab but the details in the combination form
| Possible impacts? | ~


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/27353)
<!-- Reviewable:end -->

### Breaking changes

- the suppliers association is not editable anymore from the combination form, only the product suppliers' details are The association is ALWAYS handled in the product form in Options tab
- JS component `ProductSuppliersManager` was split into two `ProductSuppliersCollection` `SuppliersSelector`
- Product mapping for product drops the redundant prefix `product.` in all fields names
- query result DTO `ProductSupplierInfo` has been removed, its name property is now in `ProductSupplierForEditing` and this DTO replaces the later (previously it was contained in it, which added an overkill layer of data)
- `SetCombinationDefaultSupplierCommand` command has been removed, only `SetProductDefaultSupplierCommand` can be used now
- `SetCombinationSuppliersCommand` is replaced by `UpdateCombinationSuppliersCommand` which can only update existing associations
- `ProductSupplierOptions` now contains the list of associated suppliers (via `SupplierId`) regardless of the product suppliers' details
- `RemoveAllAssociatedCombinationSuppliers` no longer exist
- all suppliers associations are now handled via `SetSuppliersCommand` the same command is used regardless of the product type (combinations or not) You MUST set an association before you can update its details
- `SetProductSuppliersCommand` was replaced by `UpdateProductSuppliersCommand` which can only be used to update existing associations
- `GetSupplierForViewingHandler` constructor changes, it is now based on a currencyId and the context is not used for injection anymore which makes it more stable
- class `PrestaShop\PrestaShop\Core\Domain\Product\Supplier\ProductSupplier` was renamed into `ProductSupplierUpdate` to reduce confusion it is now based on a new `SupplierAssociationInterface` VO which identifies a product supplier association via its relations' IDs (and option productSupplierId)

jolelievre committed:
Extract product suppliers edition in separated form type

jolelievre committed:
Adapt form theme to new collection type

jolelievre committed:
Remove useless form theme

jolelievre committed:
Adapt ProductFormDataProvider to new suppliers data format

jolelievre committed:
Migrate product supplier code and ProductFormModel to TypeScript

jolelievre committed:
Create dedicated types for suppliers

jolelievre committed:
Create dedicated component for suppliers selection with callback to sync selected and default suppliers

jolelievre committed:
Create dedicated component to handle product suppliers collection with setter to sync with selectors when needed

jolelievre committed:
Convert combination and product js into typescript

jolelievre committed:
Introduce dedicated combination mapping and update product map for suppliers

jolelievre committed:
Remove suppliers manager which has been split into two dedicated components

jolelievre committed:
Update necesary JSDoc for TypeScript compatibility

jolelievre committed:
Add data attribute to get the index properly

jolelievre committed:
Create a command SetSuppliersCommand to associate suppliers to a product

jolelievre committed:
Fix bugs in methods to get default supplier, associations and useless suppliers

jolelievre committed:
Set default currency when associating suppliers

jolelievre committed:
Add behat tests for suppliers association command

jolelievre committed:
Use product type instead of associations to trigger invalid type exception

jolelievre committed:
Set default supplier is authorized regardless of the type now

jolelievre committed:
Add tests on assocation command to make sure it does not change the existing data

jolelievre committed:
Rename SetProductSuppliersCommand into UpdateProductSuppliersCommand

jolelievre committed:
ProductSupplier DTO is turned into a dedicated Update DTO, it uses ProductSupplierAssociation VO which can contain the whole association parameters with optional productSupplierId

jolelievre committed:
Abstract handler now loads based on the association not just the productSupplierId

jolelievre committed:
Now product suppliers update command is only dedicated to update no more association

jolelievre committed:
Update product suppliers behat test, association must be done before the update now

jolelievre committed:
Introduce interface for product supplier association which has two implementing classes one for product and one for combination

jolelievre committed:
Simplify abstract exception creation thanks to __toString method

jolelievre committed:
ProductSupplierUpdate is based on association interface

jolelievre committed:
Adapt handlers to new association interface

jolelievre committed:
Clean code in updater and fix bugs related to new implentation from repository

jolelievre committed:
Improve product suppliers options query, it now always includes supplier IDs, informations only when not combination type and it does not fetch the whole product anymore

jolelievre committed:
Throw invalid exception when association is not found during update

jolelievre committed:
Improve suppliers behat tests to check supplier IDs instead of product suppliers content, also association can now be checked independently

jolelievre committed:
Update combination behat tests to handle all the new features, associate before updating, invalid association checking, no detailed data for options in combination products

jolelievre committed:
Rename command to update combination suppliers along with related methods and behat steps

jolelievre committed:
Improve suppliers update interface to returned list of updated associations

jolelievre committed:
Improve behat tests to use returned updated associations, and combinations check suppliers

jolelievre committed:
Update and clean all the combination tests

jolelievre committed:
Set suppliers command returns the list of all associations, recently created and existent

jolelievre committed:
Product suppliers references are created during the association command, applied for product tests only for now

jolelievre committed:
Product supplier reference is now assigned only during the association command test, they are asserted in other tests

jolelievre committed:
Test use case when we update product supplier without their IDs

jolelievre committed:
Add steps that simply check the associated suppliers

jolelievre committed:
We can define product suppliers reference when we associate suppliers to a product with combinations, then these associations can be asserted in combination behat tests

jolelievre committed:
Add behat test to update combination suppliers without specifying productSupplierId

jolelievre committed:
Add test when product default supplier is updated, wholesale price is also updated

jolelievre committed:
Fix update command test without productSupplierId reference

jolelievre committed:
Remove SetCombinationDefaultSupplierCommand

jolelievre committed:
Remove RemoveAllAssociatedCombinationSuppliersCommand

jolelievre committed:
RemoveAllAssociatedProductSuppliersCommand is usable for product with combinations since it is the only one left for all types of products

jolelievre committed:
Optimize combination suppliers update handler, no need for the whole product only its ID is needed

jolelievre committed:
Create ProductSupplierRepository::getByAssociation which simplifies handlers and updater a lot

jolelievre committed:
Handle synchronization between product supplier update and combination wholesale price and supplier reference

jolelievre committed:
Remove useless function

jolelievre committed:
Update combination wholesale price updates its associated default supplier price

jolelievre committed:
When new combinations are created the associated suppliers are associated to new combinations

jolelievre committed:
Always create an association for product with combinations even when no combination are present

jolelievre committed:
Fix bug of supplier viewing handler when currency context is not set, use configured default currency instead

jolelievre committed:
Improve tests regarding product of type combinations which have no combinations

jolelievre committed:
Revert product suppliers once again and reintroduce the association with id_product_attribute 0 for products with combinations, they are actually useful to keep the associations even when the combinations are not present

jolelievre committed:
Add tests whan product type is changed

jolelievre committed:
Remove dump

jolelievre committed:
Adapt product suppliers command builder to new format

jolelievre committed:
Improve row visibility toggle

jolelievre committed:
Fix product form data provider so that associations are always returned even without the details

jolelievre committed:
Adat product type listener so that suppliers are never removed, but product_suppleirs can

jolelievre committed:
Adapt combination supliers builde to new format

jolelievre committed:
Integrate notion of NoSupplierId

jolelievre committed:
Add new GetAssociatedSuppliers query more reduced than options query

jolelievre committed:
Use new GetAssociatedSuppliers in behat tests

jolelievre committed:
Adapt combination form data to new format

jolelievre committed:
Handle new format form combination in form type

jolelievre committed:
Fix duplicate behat tests

jolelievre committed:
We no longer use redundant product prefix in all product mappings

jolelievre committed:
ProductSuppliersCollection is now configured with a defaultSupplierId so that it can work stand alone without a reference to the product object model

jolelievre committed:
Update responsibilities, it is now the product suppliers collection which has a callback for default values since these values are only available in the collection anyway

jolelievre committed:
Default supplier ID is now handled directly via the form object model

jolelievre committed:
Init collection component with initial wholesalePrice to inject it during first creations

jolelievre committed:
Update function name in suppliers collection component

jolelievre committed:
Add error message when component is not correctly initialized

jolelievre committed:
Move combination JS code in a more conventional folder, init the CombinationFormModel component and integrate the JS code for suppliers collection

jolelievre committed:
ProductSupplierForEditing can store all data including supplier name which makes ProductSupplierInfo now useless

jolelievre committed:
Clean every references to ProductSupplierInfo and remove it

NeOMakinG reviewed (dismissed):
Honestly I couldn't find something to say, look fine enough to me for the front-end part

jolelievre commented:
@NeOMakinG can you re-approve please? I resolved the conflicts

jolelievre committed:
Missing use statement

matks reviewed (dismissed):
That was a very big PR.

There was a few things we could improve, some variable names, some little pieces of logic but nothing worth blocking this PR and it can always be improved later. Overall logic seems OK, tests seem OK. I hope I did not miss anything in that big contribution 😅

jolelievre committed:
Fix little type compilation error

jolelievre committed:
Update fixtures wholesale price so that they are consistent with suppliers data

jolelievre commented:
Hi @matks and @NeOMakinG 
I need re-approval for this PR after rebase to fix conflicts

khouloudbelguith reviewed (dismissed):
Hi @jolelievre,

**Issue1**
1. Enable Experimental product page
2. Create a product with combinations V2 (cost price different from 0) 
![image](https://user-images.githubusercontent.com/16067358/154998797-1264345c-a82d-4ebf-8507-6258219d3a73.png)
3. Select a supplier for this product 
![image](https://user-images.githubusercontent.com/16067358/154999025-38ae7fd3-9aa7-4e9c-a5c2-535f0bb9f2ef.png)
4. Edit one combination & set the cost price = 3 & save 
![image](https://user-images.githubusercontent.com/16067358/154999200-3559afaf-e814-40f8-886a-d465376d347e.png)
The other combination 
![image](https://user-images.githubusercontent.com/16067358/154999234-b6f7a5e0-21eb-4e40-ab0c-8eabf2901e4d.png)
5. Go to FO > Order two combinations & place the order 
![image](https://user-images.githubusercontent.com/16067358/154999418-66f5b3b3-eecc-453f-9011-744f22ee28dc.png)
6. Check the table PREFIX_order & check the column `purchase_supplier_price` & `original_wholesale_price`
![image](https://user-images.githubusercontent.com/16067358/154999879-fdfcac51-7ecc-4f19-9ba5-dbe770d5b90b.png)

It seems not correct because the second combination has no cost price

**Issue2**
1. Enable Experimental product page
2. Enable multistore & create a new shop
3. Create a product in all shops with V1
4. Select a shop1 & try to edit the product with V2 => NOK 
![image](https://user-images.githubusercontent.com/16067358/155000429-8b8bbf63-bc4e-4556-9e64-e18e42938a02.png)

I attached a screen record

https://user-images.githubusercontent.com/16067358/155000765-46a560c6-919e-4f6a-b8fb-c93bd8f7fdce.mp4

**without PR** => ok
I attached a screen record


https://user-images.githubusercontent.com/16067358/155001113-3adeabb2-5327-49c7-9d82-11213ceec57c.mp4

Thanks!

jolelievre committed:
Fix bug when default supplier is updated on product with combinations that has no combinations yet

jolelievre committed:
Handle product wholesale price remains independent from combination updates, and we do not remove product supplier associations on combination supplier updates

jolelievre commented:
Hi @khouloudbelguith 

thanks for your feedback

#### Issue 1
There was some inconsistent logic related to your issue:
- first when combination has no wholesale price it means it doesn't override it, so the product's wholesale price is used
- there was a bug that always updated the product's wholesale price whenever a combination was updated

So when you changed the value for Combination1 it was applied to the product, and then this value was used as a fallback for Combination2.

So I fixed the problem this way: for product with combinations the Product's wholesale price remains independent from the updates in combinations This way it can be used as the default value for all combinations which wholesale_price remains zero, but if they are overridden then the combination's value is used.

I also fixed a bug that I didn't notice when all combinations were removed.

#### Issue 2
This one is trickier because it's related to the new product's type field. So first it happens regardless of the multishop feature, you can simply create a product in legacy page switch to the new one and you'll see the error. The problem is that we still allow product to be created without the type which should be mandatory.

The bug was already present, it was just invisible. This PR does highlight it because we now fetch and check the product type more strictly but it's a side effect of this PR.

I could make a quick fix for this one but it would hide lots of other use cases that would pop up unexpectedly, that's why I won't fix it in this PR. Instead I prefer to merge the code as is and open a dedicated issue for this bug because we'll have to look in details all the references to this product type and correctly fix them or handle a proper fallback (for products that would be created in old versions, or imported without this type).

So could you please open a dedicated issue with the buggy scenario you found? (no need for multishop)

jolelievre committed:
Disabled advacned stock in tests even though it is not needed the CI needs it for some reason

Progi1984 commented on src/Adapter/Product/AbstractProductSupplierHandler.php:
```diff
@@ -58,57 +57,42 @@ public function __construct(
      * @return ProductSupplier
      */
-    protected function loadEntityFromDTO(ProductId $productId, ProductSupplierDTO $productSupplierDTO, ?CombinationId $combinationId = null): ProductSupplier
+    protected function loadEntityFromDTO(ProductSupplierUpdate $productSupplierUpdate): ProductSupplier
```
Please document the BC Break

Progi1984 commented on src/Adapter/Product/Combination/CommandHandler/GenerateProductCombinationsHandler.php:
```diff
@@ -42,20 +43,31 @@ final class GenerateProductCombinationsHandler implements GenerateProductCombina
     public function __construct(
-        CombinationCreator $combinationCreator
+        CombinationCreator $combinationCreator,
+        ProductSupplierUpdater $productSupplierUpdater
```
Please document the BC Break

Progi1984 replied:
It's not a BC Break as it's a final

Progi1984 commented on src/Adapter/Product/Combination/CommandHandler/SetCombinationDefaultSupplierHandler.php:
```diff
@@ -1,73 +0,0 @@
-use PrestaShop\PrestaShop\Core\Domain\Product\Combination\CommandHandler\SetCombinationDefaultSupplierHandlerInterface;
-use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
-
-class SetCombinationDefaultSupplierHandler implements SetCombinationDefaultSupplierHandlerInterface
```
Please define the BC Break

Progi1984 commented on src/Adapter/Product/Combination/CommandHandler/UpdateCombinationPricesHandler.php:
```diff
@@ -29,27 +29,40 @@
-    {
+    public function __construct(
+        CombinationRepository $combinationRepository,
+        ProductSupplierRepository $productSupplierRepository
```
Not a BC Break, because previously the class was final.

Progi1984 commented on src/Adapter/Product/Supplier/CommandHandler/SetSuppliersHandler.php:
```diff
@@ -23,18 +23,19 @@
- * Removes product suppliers associated to a certain combination using legacy object model
+ * Handles @see SetSuppliersCommand using legacy object model
  */
-final class RemoveAllAssociatedCombinationSuppliersHandler implements RemoveAllAssociatedCombinationSuppliersHandlerInterface
```
Please document the BC Break ?

Progi1984 commented on src/Adapter/Product/Supplier/CommandHandler/UpdateProductSuppliersHandler.php:
```diff
@@ -31,13 +31,13 @@
- * Handles @see SetProductSuppliersCommand using legacy object model
+ * Handles @see UpdateProductSuppliersCommand using legacy object model
  */
-final class SetProductSuppliersHandler extends AbstractProductSupplierHandler implements SetProductSuppliersHandlerInterface
```
Please document the BC Break ?

Progi1984 reviewed (commented):
@jolelievre Some feedback :)

jolelievre commented:
Hi @Progi1984 @matks @NeOMakinG 

The PR has been rebased and needs a re-approval please

khouloudbelguith reviewed (dismissed):
Hi @jolelievre,

Could you please check the Conflicting files?

Thanks!

Amoifr commented on src/Adapter/Product/AbstractProductSupplierHandler.php:
```diff
@@ -58,57 +57,42 @@ public function __construct(
      * @return ProductSupplier
      */
-    protected function loadEntityFromDTO(ProductId $productId, ProductSupplierDTO $productSupplierDTO, ?CombinationId $combinationId = null): ProductSupplier
+    protected function loadEntityFromDTO(ProductSupplierUpdate $productSupplierUpdate): ProductSupplier
```
could you typehint your params ?

jolelievre replied:
I don't understand, they are already typed 🤔

Amoifr replied:
yes I'm wrong sorry

Amoifr commented on src/Adapter/Product/Combination/CommandHandler/UpdateCombinationPricesHandler.php:
```diff
@@ -60,6 +73,31 @@ public function handle(UpdateCombinationPricesCommand $command): void
+    {
+        $productId = $this->combinationRepository->getProductId($combinationId);
+        $defaultSupplierId = $this->productSupplierRepository->getDefaultSupplierId($productId);
+        if (!$defaultSupplierId) {
```
!$defaultSupplierId does it means === 0 or ===null ? I always prefer explicit condition. ;-)

jolelievre replied:
`getDefaultSupplierId` returns `SupplierID|null` and is strongly typed, so it can only be null anyway

But I understand the need for explicit condition ^^

Amoifr commented on src/Adapter/Product/Combination/CommandHandler/UpdateCombinationPricesHandler.php:
```diff
@@ -60,6 +73,31 @@ public function handle(UpdateCombinationPricesCommand $command): void
+        }
+
+        $defaultProductSupplier = $this->productSupplierRepository->get($defaultProductSupplierId);
+        $defaultProductSupplier->product_supplier_price_te = (float) (string) $wholesalePrice;
```
I think the product_supplier_price_te came from legacy and we are not able to give an explicit name instead "te" ?

jolelievre replied:
That's the legacy name indeed so we won't change it to avoid a breaking change

At least the field explicitly indicates that the value is tax excluded (te = tax excluded), in most of the DB prices don't have a suffix but they are mostly tax excluded which must be know by convention 🤷‍♂️. So it's not that bad 😅

Amoifr commented on src/Adapter/Product/Combination/Repository/CombinationRepository.php:
```diff
@@ -106,6 +106,23 @@ public function get(CombinationId $combinationId): Combination
         return $combination;
     }
 
+    public function getProductId(CombinationId $combinationId): ProductId
```
```suggestion
    public function getProductIdByCombinationId(CombinationId $combinationId): ProductId
```

jolelievre replied:
I think it's quite explicit since we are in the `CombinationRepository` and, most of all, there is only one parameter which is typed with a VO ProductId, if it was a simple int I would agree that it needs more explicit naming but here the typing says it all

Just like the following function `getDefaultCombinationId` doesn't need to be named `getDefaultCombinationIdByProductId` it is a but redundant

Amoifr replied:
different habits .... but I guess to know how you will name a future method that get the product id by an another dimension. 
getProductId(FooId $fooId) ?

Amoifr commented on src/Adapter/Product/QueryHandler/GetProductSupplierOptionsHandler.php:
```diff
@@ -64,11 +66,20 @@ public function __construct(
-            (int) $product->id_supplier,
-            $this->getProductSuppliersInfo($query->getProductId())
+            null !== $defaultSupplier ? $defaultSupplier->getValue() : 0,
+            array_map(function (SupplierId $supplierId) {
```
I'm not very friendly with making operation in method call parameters. On my point of view it's not easily readeble

jolelievre replied:
I don't mind personally, but ok!

Amoifr commented on src/Adapter/Product/Repository/ProductSupplierRepository.php:
```diff
@@ -107,18 +187,19 @@ public function getDefaultSupplierId(ProductId $productId): ?SupplierId
-            ->innerJoin(
-                'ps',
-                $this->dbPrefix . 'product',
+            ->from($this->dbPrefix . 'product', 'p')
```
we are in the ProductSupplierRepository custom dictates that we request on this entity first in the from and left join others ?

jolelievre replied:
Yeah, but the data `p.id_supplier` representing the default supplier ID comes from the product table, that's why I built the request this way, regardless of the class that performs the request

The domain of the repository is ProductSupplier but the data must be fetched from product table, but it ensures there is a correct mapping in the product_supplier table thanks to the rightJoin This is used to detect if the data in product table is still valid or needs to be updated

When the request doesn't match with an association it returns null, so we know the default supplier must be updated, this can happen when a ProductSupplier has been removed

Amoifr commented on src/Adapter/Product/Repository/ProductSupplierRepository.php:
```diff
@@ -163,27 +244,59 @@ public function getDefaultProductSupplierId(ProductId $productId): ?ProductSuppl
 
-        return array_map(function (array $result) {
-            return new ProductSupplierId((int) $result['product_supplier_id']);
+        return array_map(function (array $row) {
```
```suggestion
        return array_map(static function (array $row): SupplierId {
```

Amoifr commented on src/Adapter/Product/Repository/ProductSupplierRepository.php:
```diff
@@ -276,10 +389,42 @@ public function getProductSuppliersInfo(ProductId $productId, ?CombinationId $co
+            return [];
+        }
+
+        return array_map(function (array $row) {
```
```suggestion
        return array_map(static function (array $row): ProductSupplierId {
```

Amoifr commented on src/Adapter/Product/Supplier/QueryHandler/GetAssociatedSuppliersHandler.php:
```diff
@@ -0,0 +1,69 @@
+
+        return new AssociatedSuppliers(
+            $defaultSupplier ? $defaultSupplier->getValue() : NoSupplierId::NO_SUPPLIER_ID,
+            array_map(function (SupplierId $supplierId) {
```
```suggestion
            array_map(static function (SupplierId $supplierId): int {
```

Amoifr commented on src/Adapter/Product/Update/ProductSupplierUpdater.php:
```diff
@@ -136,93 +242,128 @@ public function removeAllForProduct(ProductId $productId): void
-        $this->updateDefaultSupplier($productId, $supplierId, $combinationId);
+        $combination = $this->combinationRepository->get(new CombinationId((int) $defaultCombinationSupplier->id_product_attribute));
+        $combination->supplier_reference = $defaultCombinationSupplier->product_supplier_reference;
+        $combination->wholesale_price = (float) (string) $defaultCombinationSupplier->product_supplier_price_te;
```
I had never seen this before ;-)
(float) (string)  we do not have any mathod or service that will do that explicitly

jolelievre replied:
That's a typo, probably a copy-paste from a code using `DecimalNumber`, that's how we convert `DecimalNumber` to float usually, but here it's needed since the data comes from a legacy object

Amoifr commented on src/Adapter/Product/Update/ProductSupplierUpdater.php:
```diff
@@ -136,93 +242,128 @@ public function removeAllForProduct(ProductId $productId): void
-            throw new ProductSupplierNotAssociatedException(sprintf(
-                'Supplier #%d is not associated with product #%d', $supplierIdValue, $productIdValue
-            ));
+        $product = $this->productRepository->get(new ProductId((int) $defaultProductSupplier->id_product));
```
Funny ! we construct a ProductId here, and the first step of the get method is to do a getValue on it :-)

jolelievre replied:
Indeed, but since the prototype of `ProductRepository::get` asks for a `ProductId` parameter we don't have a choice. This is to "strictify" the methods from repositories by using appropriate VOs as much as possible

Amoifr commented on src/Core/Domain/Product/Combination/Command/UpdateCombinationSuppliersCommand.php:
```diff
@@ -58,7 +60,7 @@ public function __construct(
     ) {
         $this->combinationId = new CombinationId($combinationId);
-        $this->setCombinationSuppliers($combinationSuppliers);
+        $this->setCombinationSuppliers($combinationSuppliers, $combinationId);
```
is it necessary to add $combinationId in param you have already the property $this->combinationId set jus a line before ?

jolelievre replied:
Yep, you're right!

Amoifr commented on src/Core/Domain/Product/Combination/Command/UpdateCombinationSuppliersCommand.php:
```diff
@@ -80,23 +82,26 @@ public function getCombinationSuppliers(): array
-                $productSupplier['supplier_id'],
+            $this->combinationSuppliers[] = new ProductSupplierUpdate(
+                new CombinationSupplierAssociation(
+                    $combinationId,
```
```suggestion
                    $this->combinationId->getValue(),
```

Amoifr commented on src/Core/Domain/Product/Combination/Exception/CannotUpdateCombinationException.php:
```diff
@@ -57,4 +57,9 @@ class CannotUpdateCombinationException extends CombinationException
+    /**
+     * When fails to update default supplier data
+     */
+    public const FAILED_UPDATE_DEFAULT_SUPPLIER_DATA = 60;
```
is it use as errorCode on an exception ? if the respose is yes, it's ok. If the response is no, why using meaningless numbers

jolelievre replied:
Yep these are all error codes, so the main importance is that they are all different to identify each error case

Amoifr commented on src/Core/Domain/Product/Supplier/Exception/InvalidProductSupplierAssociationException.php:
```diff
@@ -21,21 +22,12 @@
-{% endblock %}
+class InvalidProductSupplierAssociationException extends ProductSupplierException
+{
+}
```
empty ! normal .... nothing more than its parent ?

jolelievre replied:
No special information is required for this exception, but we need a specific type/class to catch the appropriate exception if needed

Amoifr replied:
ok thx

Amoifr commented on src/Core/Domain/Product/Supplier/ProductSupplierUpdate.php:
```diff
@@ -29,18 +29,17 @@
+     * @var SupplierAssociationInterface
      */
-    private $supplierId;
+    private $association;
```
can we be more verbose and use 
```suggestion
    private $supplierAssociation;
```

jolelievre replied:
I made it match the getter `getAssociation` and I named it a bit shorter on purpose because this is in the `ProductSupplier` namespace. In the repository, I also have a `ProductSupplierRepository::getByAssociation` which uses the same short naming

I think the namespace + the strict type everywhere is explicit enough 🤷‍♂️

You think using just `association` is confusing?

Amoifr commented on src/Core/Domain/Product/Supplier/QueryResult/AssociatedSuppliers.php:
```diff
@@ -26,46 +26,43 @@
+     * @var int[]
      */
-    private $defaultSupplierId;
+    private $supplierIds;
```
it allways hurts me we we lose informations, and where a getSupplierIds return an array of int instead of an array of SupplierId

jolelievre replied:
That's because this is a QueryResult our convention is to rely on flattened/native types for QueryResult because we master its creation unlike command parameters which come from outside and needs to be validated/asserted

Amoifr commented on src/Core/Form/IdentifiableObject/CommandBuilder/Product/ProductSuppliersCommandsBuilder.php:
```diff
@@ -40,36 +41,45 @@ final class ProductSuppliersCommandsBuilder implements ProductCommandsBuilderInt
     public function buildCommands(ProductId $productId, array $formData): array
     {
-        if (!isset($formData['options']['suppliers']['product_suppliers']) && !isset($formData['options']['suppliers']['default_supplier_id'])) {
+        if (!isset($formData['options']['suppliers']['supplier_ids']) &&
```
&& or || do we need all or a least missing one to avoid the action ?

jolelievre replied:
It's a && we need all of them to be empty which means no data is present for either of the three commands handled by this builder

Then in the lower code, each command checks for its respective data individually

Amoifr replied:
Are you agree to say that this expression can be difficult to understand. In clean code is recommanded to avoid multiple conditions especially when they use the not option.
For the future reader may be we can move into a private method with an explicit name ?

Amoifr commented on src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php:
```diff
@@ -642,36 +642,41 @@ private function presentTags(array $localizedTagsList): array
      * @param ProductForEditing $productForEditing
      *
-     * @return array<string, int|array<int, int|array<string, string|int>>>
+     * @return array<string, array<int|string, array<int|string, int|string|null>|int>>
```
Do you think it's welcome to create a DTO that replicate this structure to simplify ?

Amoifr replied:
it will be also easier to access the data with a getter than list of keys.

jolelievre replied:
All our form data rely on array data, we don't use DTO in our forms This allows extending the data more easily because a module can add a form type in any part of the whole form, and add the associated data in the array

With full DTO it would imply extending the DTO class to add fields, which makes it harder especially if multiple modules want to override the same form type

jolelievre reviewed (commented):
Thanks @Amoifr 

I handled most of your feedback or answered them when I didn't fully agree (I'm open to discussion)

jolelievre committed:
Handle feedback from review

Progi1984 reviewed (dismissed):
Hello @jolelievre, thanks for your contribution, would you mind to rebase your pull request? Thanks

Reference : https://anavarre.net/how-to-rebase-a-github-pull-request/

eternoendless commented on src/Core/Domain/Product/Supplier/Command/SetSuppliersCommand.php:
```diff
@@ -0,0 +1,88 @@
+use PrestaShop\PrestaShop\Core\Domain\Supplier\ValueObject\SupplierId;
+use PrestaShop\PrestaShop\Core\Exception\InvalidArgumentException;
+
+class SetSuppliersCommand
```
What is this command for? --> document it

eternoendless commented on src/PrestaShopBundle/Form/Admin/Type/TranslatorAwareTypeTrait.php:
```diff
@@ -0,0 +1,89 @@
+ * to extend a specific type (CollectionType, ChoiceType, ...) it is convenient to have these features available
+ * in this trait (which is also used by TranslatorAwareType).
+ */
+trait TranslatorAwareTypeTrait
```
I have the feeling that we will regret doing this

eternoendless commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Options/ProductSupplierCollectionType.php:
```diff
@@ -0,0 +1,70 @@
+    {
+        parent::configureOptions($resolver);
+        $resolver->setDefaults([
+            'label' => $this->trans('Supplier reference(s)', 'Admin.Catalog.Feature'),
```
I don't think it's not worth creating a trait just to use `$this->trans()` instead of `$this->translator->trans()`

eternoendless commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Options/ProductSupplierCollectionType.php:
```diff
@@ -0,0 +1,70 @@
+    public function __construct(TranslatorInterface $translator, array $locales)
+    {
+        $this->translator = $translator;
+        $this->locales = $locales;
```
This doesn't look used. Are you passing this parameter just because the trait requires so?

eternoendless commented on src/Core/Domain/Product/Supplier/ProductSupplierUpdate.php:
```diff
@@ -29,18 +29,17 @@
+use PrestaShop\PrestaShop\Core\Domain\Product\Supplier\ValueObject\SupplierAssociationInterface;
 
 /**
  * Transfers data of product supplier
```
I think this description should be more precise

eternoendless commented on src/Core/Domain/Product/Supplier/Query/GetAssociatedSuppliers.php:
```diff
@@ -24,17 +24,32 @@
-interface RemoveAllAssociatedCombinationSuppliersHandlerInterface
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+
+class GetAssociatedSuppliers
```
Needs a description

eternoendless commented on src/Core/Domain/Product/Supplier/QueryResult/AssociatedSuppliers.php:
```diff
@@ -26,46 +26,43 @@
-use PrestaShop\PrestaShop\Core\Domain\Supplier\ValueObject\SupplierId;
-
-class SetCombinationDefaultSupplierCommand
+class AssociatedSuppliers
```
Needs a description

eternoendless commented on src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php:
```diff
@@ -575,36 +575,41 @@ private function presentTags(array $localizedTagsList): array
      * @param ProductForEditing $productForEditing
      *
-     * @return array<string, int|array<int, int|array<string, string|int>>>
+     * @return array<string, array<int|string, array<int|string, int|string|null>|int>>
```
Looks like you already know the schema of this array, consider using an array shape:
```suggestion
     * @return array{suppliers: array{default_supplier_id: int, supplier_ids: int[]}, product_suppliers: array<int, array{supplier_id: int, supplier_name: string, product_supplier_id: int, price_tax_excluded: string, reference: string, currency_id: int, combination_id: int}>}
```

jolelievre committed:
Improve PHPDoc

jolelievre committed:
Fix ProductSupplierCollectionType that relied on a removed trait

jolelievre committed:
Improve PHPDoc

jolelievre commented:
@eternoendless I handled your feedback

jolelievre committed:
Fix addSupplier which overrode the default price value always

NeOMakinG commented on admin-dev/themes/new-theme/js/pages/product/components/suppliers/product-suppliers-collection.ts:
```diff
@@ -0,0 +1,324 @@
+
+      this.productSuppliers[supplierId] = {
+        supplierId,
+        productSupplierId: <string> $(this.map.productSupplierRow.productSupplierIdInput(supplierIndex)).val(),
+        supplierName: <string> $(this.map.productSupplierRow.supplierNameInput(supplierIndex)).val(),
+        reference: <string> $(this.map.productSupplierRow.referenceInput(supplierIndex)).val(),
+        price: <number> $(this.map.productSupplierRow.priceInput(supplierIndex)).val(),
+        currencyId: <string> $(this.map.productSupplierRow.currencyIdInput(supplierIndex)).val(),
```
Not blocking, but it could be improved

Casting is not really future proof, in case something moves in the markup, it won't be detected by the JS and it will break because the element can be undefined and its type can change too...

NeOMakinG reviewed (dismissed):
I'm just a bit disappointed about these casts, but we should improve it on the global app later anyway

jolelievre commented:
@NeOMakinG this could be improved by adding some data attributes to indicate the type maybe? This could be added in a PR that handles the extension of this component maybe, because for now only the data expected from the core are persisted

NeOMakinG commented:
> @NeOMakinG this could be improved by adding some data attributes to indicate the type maybe? This could be added in a PR that handles the extension of this component maybe, because for now only the data expected from the core are persisted

No, I guess we should use type guards, parseInt etc

jolelievre commented:
@NeOMakinG ok but in this case we're fine since all the parsed data are strings anyway Which is the default type of input values no?

khouloudbelguith reviewed (commented):
Hi @jolelievre,

I found those issues:

**First issue**
1. Go to BO > Catalog > Suppliers > Create a new **disabled** supplier
2. Go to BO > Catalog > Products V2  > Create a new product with combinations, in the Options tab, check the disabled supplier is not displayed
3. Go back to BO > Catalog > Suppliers > **enable** the last supplier created
4. Go back to BO > Catalog > Products V2 > Try to edit/create a new product
5. See exception 

![image](https://user-images.githubusercontent.com/16067358/162964473-9e2ab4a6-9519-40a5-87d8-6337cdd69957.png)

https://user-images.githubusercontent.com/16067358/162964500-a06f4722-60f2-45b9-9ee0-ec15bf99e8d9.mp4

------------------------------------------------------------------

**Second issue**
1. Go to BO > Catalog > Products V2
2. Create a product with combinations
3. Generate combinations for this product & save
4. In the Options tab > Select any supplier and don't save
5. In the Combinations tab > Edit any combinations > the new supplier is not displayed

In fact, we need to save after selecting Or changing the default combinations Or unselecting a supplier in the options Tab, it is an ajax issue, I think.

I attached a screen record

https://user-images.githubusercontent.com/16067358/162965738-c0eccbb9-3562-4aa2-a9b4-9f4547905e4f.mp4

------------------------------------------------------------

**Third issue**
1. Go to BO > Catalog > Products V2
2. Create a product with combinations **enabled**
3. In the Options tab > select two supplier 
![image](https://user-images.githubusercontent.com/16067358/162966030-e594a39d-6373-4c0c-88a9-830aac6120e4.png)
4. Generate combinations & add a supplier reference for each supplier 
![image](https://user-images.githubusercontent.com/16067358/162966381-581f155a-f51e-465b-9c05-387c126d0753.png)
5. Now delete the supplier by default (for eg: Fashion supplier)
6. Go to BO > Edit the last product created & check **without saving**:
6.1 In the options tab => the first supplier is checked 
![image](https://user-images.githubusercontent.com/16067358/162967132-7a7e907d-25b6-4e16-99aa-433a64c0caa9.png)
6.2 In the combinations tab > Edit the combinations => the first supplier is well displayed  
![image](https://user-images.githubusercontent.com/16067358/162967286-0dfc1086-78f7-46c6-9b56-fe480c22643f.png)
7. Create an order with this product
8. In the BO > Orders > View the last order
9. See error: the reference of the second supplier is not displayed 
![image](https://user-images.githubusercontent.com/16067358/162969307-f46b9f87-5271-4583-9bad-42333b06e1e9.png)

It should be 
![image](https://user-images.githubusercontent.com/16067358/162969506-f38f511c-f97b-4ab5-b5b9-3aa9bdf96635.png)

Thanks!

jolelievre committed:
Fix supplier removal, update product suppleirs association and default data

jolelievre commented:
Hi @khouloudbelguith 

thanks for the feedback!

#### First issue
I wasn't able to reproduce the first issue after a fresh install and doing the exact same steps. I rebased right before I tested it so maybe just this was enough to fix the issue. Can you please check if you still have the bug? If sot then some steps might be missing in the bug description.

#### Second issue
This is not a bug it is the expected behaviour, the suppliers association is handled in the product's form where you select the suppliers. If you don't save this form the association is not created so there's no way to display it in the combination form which is independent. You have to save the product's from to validate the new associations before you can edit combination's suppliers.

#### Third issue
I reproduced it, you mentioned you didn't save so it's normal that the data is not updated. When you save on purpose everything returns fine (again the suppliers' association is not done in ajax so nothing is updated until the product form is submitted)

The problem here happens in the supplier's page, the removal doesn't handle updating products. It should be handled in a dedicated issue, but I fixed it here nevertheless. It should be fine now.

khouloudbelguith reviewed (commented):
Hi @jolelievre,

Thanks for your feedback.
>First issue
I wasn't able to reproduce the first issue after a fresh install and doing the exact same steps. I rebased right before I tested it so maybe just this was enough to fix the issue. Can you please check if you still have the bug? If sot then some steps might be missing in the bug description.

I manage to reproduce the issue with the PR, new rebase, new fresh install
I'm using php8.0,
Here's my configuration 
![image](https://user-images.githubusercontent.com/16067358/163422577-bc1d2171-fa93-4cf2-8ad5-c6c57649e886.png)

I attached a screen record

https://user-images.githubusercontent.com/16067358/163422818-ff8d1a77-7ce0-46fc-b342-1aff1078d7ce.mp4

I tried with php7.4 => same exception
Thanks!

jolelievre committed:
Remove useless choice attribute that as not configured the right way to begin with

jolelievre commented:
Hi @khouloudbelguith 

I finally managed to reproduce the bug!!! Actually, to reproduce it we needed a supplier whose name is `disabled` The form type was badly configured but the bug only triggered when a supplier matching this name existed It's fixed now!

jolelievre committed:
Add constraint on supplier price

jolelievre committed:
Add constraint in combination form

jolelievre committed:
Fix query to be compatible with mariadb

jolelievre committed:
Fix cache issue that messed with categories association and tag tests

HanaRebaiQA reviewed (commented):
Hello @jolelievre 

Here's the encountered issues while testing multiple times.


**First issue** : An exception is thrown when putting a negative amount in Cost price field

There are two problems here: 

- The wholesale price is not updated when changing the Cost price in Supplier field
- An alert is missing to notify that the price should be positive or zero (as without PR)

1. BO > Catalog > Products (experimental) V2  > Create a product
2. In `Supplier references` section > Add a negative amount in Cost price field 
3. Click on Save
4. See error > Exception

![image](https://user-images.githubusercontent.com/92912932/165497563-a9a00281-8da5-4057-bfa5-02570ac23c99.png)

_With PR_

https://user-images.githubusercontent.com/92912932/165497632-d7d0c155-da1b-4979-9280-d43071b76fb3.mp4

_Without PR_ 

https://user-images.githubusercontent.com/92912932/165498455-098782ea-aed8-4ff5-918e-20aebd553dc0.mp4

==> This issue is verified when tested it again following it's fix. To be tested again in the next check.

**Second issue** : Webservices issues

There are two issues when create a product V2 with **API** 

**_Issue 1_ :** An exception is thrown when edit a product V2 which is created with API


1. Create a product with API
5. The product is displayed in the product page V2 
6. Click to edit the product 
7. See error > Exception

![image](https://user-images.githubusercontent.com/92912932/165502155-b8d003ed-63a2-41b1-a5ae-2ea3fe0b54a2.png)

https://user-images.githubusercontent.com/92912932/165502229-52d22f04-4831-4c7d-ba90-59dd87b11554.mp4

_Without PR :_ 

https://user-images.githubusercontent.com/92912932/165502298-72d87c09-c7c5-4342-8d26-126c42339c7a.mp4

**==> The issue 1 is not present without PR**


**_Issue 2_ :** 


1. Create a product with API
2. The product is displayed in the product page V2 
3. Change the XML object by adding the product type (**Only with this PR to can edit the product**)

`<product_type>combinations</product_type>
`

4. Click to edit the product 
5. Click on Save
6. See error > Exception

![image](https://user-images.githubusercontent.com/92912932/165503706-6950c76a-006d-4c29-a8fd-e3c2eb21b7d2.png)

https://user-images.githubusercontent.com/92912932/165503769-2b93ed53-a19c-44ed-b1cd-0edfccf52136.mp4

**==> The issue 2 is present without PR**


**The impact of this second issue** : The product images are not well displayed following this issue. 

![image](https://user-images.githubusercontent.com/92912932/165504381-c358a059-e7a6-44e6-a43d-50b731ee6a1e.png)


@MatShir : These two issues could be fixed in this PR or i create a dedicated issue ? what do you think ?


**Third issue** : The supplier still displayed in` Supplier references` after being diabled

1. BO > Catalog > Brands & Suppliers > Suppliers > Create a supplier and enabled it
2. Configure this supplier in product V2 with one of these product types (Standard, Virtual, Pack) and Save
3. Go Back to `Suppliers` page and disable the created supplier
5. Come back to product page > Edit the product which has this supplier
6. See error > The supplier still displayed in the `Supplier references` section

![image](https://user-images.githubusercontent.com/92912932/165504869-6d663aa3-1980-480b-993d-1039cd450b29.png)


**Fourth issue** : An exception is thrown when save a product which contains a disabled supplier

This issue is related the Third one.

1. BO > Catalog > Brands & Suppliers > Suppliers > Create a supplier and enabled it
2. Configure this supplier in product V2 with one of these product types (Standard, Virtual, Pack) and Save
3. Go Back to `Suppliers` page and disable the created supplier
4. Come back to product page > Edit the product which has this supplier
6. Save the product 
7. See error > Exception 

![image](https://user-images.githubusercontent.com/92912932/165514163-bb5d6c37-a70e-4dad-826d-858997afd2cf.png)

https://user-images.githubusercontent.com/92912932/165509247-f23733be-350c-414a-bcb5-0756261e1f56.mp4



**Fifth issue** : The currency symbol in Cost price field is not updated when change the currency

1. BO > IMPROVE > International > Localization > Configure a new localization pack
2. In BO > Catalog > Products (experimental) > Edit a product V2
3. Use any product type
4. In Supplier references section > Change the currency
5. See error > The currency symbol in Cost price field is not updated 

![image](https://user-images.githubusercontent.com/92912932/165507380-d69293c0-3bd8-4cad-ae63-c68109736ba8.png)

https://user-images.githubusercontent.com/92912932/165508312-6aa4f731-d1c6-47cd-a0ff-01e318b514c8.mp4

I also have created an [issue](https://github.com/PrestaShop/PrestaShop/issues/28338) related to Product V2 but it's not related to this PR.

**PS:** Could you please check the Conflicting files ?

Thank you!

jolelievre committed:
Add scenario when product supplier is disabled

jolelievre committed:
Use all suppliers in form type even disabled ones

jolelievre committed:
Dynamically update the money type symbol for supplier prices

jolelievre committed:
Fix import because file was moved

jolelievre committed:
Clear cache ProductResetter

MatShir commented:
- First Issue and Second issue should be fixed in a different PR and on a different issue. 
- Third issue and Fourth issue is fixed by displaying all the supplier, even the disabled one (the specs will be updated, first thing on Monday)
- Fifth issue will be done in a separate issue since it is an improvement (bug already present in old version), I will put at the end of the back of since it is not a priority :)

khouloudbelguith reviewed (commented):
Hi @jolelievre,

**First issue**

1. Create a new currency (you should have at least two currencies) for example `DZD` 
![image](https://user-images.githubusercontent.com/16067358/166950496-dbd94c76-53ac-4819-bb83-99996dd5df2f.png)
2. Create a simple product with V2 
3. In the options tab, select a supplier and choose the second currency and save 
![image](https://user-images.githubusercontent.com/16067358/166951159-e5afa9ba-9a2a-46f2-a873-75d987c97b5b.png)
4. Create a product with combinations & for each combination, choose a supplier and select the second currency 
![image](https://user-images.githubusercontent.com/16067358/166951930-ad46d184-d909-4bbf-9ffb-3dd6f4720b48.png)
5. Go to BO > International > Localization > Currencies page 
6. Delete or disable the second currency
7. Try to edit the last product Simple product > Options Tab 
8. See error > the currency DZD is still displayed
![image](https://user-images.githubusercontent.com/16067358/166952452-49bfa7ea-98bd-4e3a-9bd4-b700cb985d44.png)
![image](https://user-images.githubusercontent.com/16067358/166952777-4f9a4693-1d12-4db0-a18b-e1b9c50a3c4f.png)
9. Try to edit the last product with combinations > edit a combination
10. See error  > the currency DZD is still displayed 
![image](https://user-images.githubusercontent.com/16067358/166955018-45b5bf43-4984-42f3-b392-ed31c4e1b436.png)

**Second issue**
1. Create a standard simple product
2. In the options tab > Choose a supplier and save
3. Go to BO > Advanced Parameters > Import page
4. Import combinations
I used this example
```php
Product ID*;Product_Reference;Attribute (Name:Type:Position)*;Value (Value:Position)*
48;reference1;Color:color:0;Blue:0
48;reference2;Color:color:0;Yellow:1
```
5. Go BO > Products page
6. Edit the last product & edit the combination
7. See error > no supplier displayed

I attached a screen record

https://user-images.githubusercontent.com/16067358/166956408-877417a3-68c1-4f08-b836-0bd1e78894e3.mp4

**Third issue**
1. Create a new currency (you should have at least two currencies) for example `DZD` 
2. Create a simple product with V2 
3. In the options tab, select at least two suppliers and choose the second currency
4. Unselect a supplier
5. See error, the currency is automatically changed to the default one

I attached a screen record

https://user-images.githubusercontent.com/16067358/166958512-d406ac63-a75c-4e3d-b1d4-100582b7cbf0.mp4

Thanks!

jolelievre committed:
Fix selector that watch inputs to memorize them, currency select was ignored

MatShir commented:
1.  The bug is already present, but couldn't be seen before. The migration of product page has reveled the bug. 
I suggest making a new issue and see how we can solve this pb :)

@khouloudbelguith is it ok for you ?

jolelievre committed:
Quick fix for suppliers with identical names

jolelievre committed:
Integrate ID in supplier name only when there are identical ones

jolelievre committed:
Fix hasDuplicateSuppliersName repository method

jolelievre committed:
Order by id after name

khouloudbelguith reviewed (approved):
Hi @jolelievre,

I checked this PR after the [last commit](https://github.com/PrestaShop/PrestaShop/pull/27353/commits/14ced7c71cade1b6c666e50587ab8138ca734734)

It is ok :heavy_check_mark: 
Automatic tests => ok :heavy_check_mark: 
But we need to create some issues
1. The exception as commented here https://github.com/PrestaShop/PrestaShop/pull/27353#pullrequestreview-888951937 **issue2**
2. The exception after adding a product via API as commented here: https://github.com/PrestaShop/PrestaShop/pull/27353#pullrequestreview-954773216 **Second issue: Issue1 & Issue2**
3. The currency symbol in Cost price field is not updated when change the currency as commented here: https://github.com/PrestaShop/PrestaShop/pull/27353#pullrequestreview-954773216 **Fifth issue** and commented here: https://github.com/PrestaShop/PrestaShop/pull/27353#pullrequestreview-963560433 **First issue** + This case 
```php

1. Create a new currency (you should have at least two currencies) for example DZD
2. Create a simple product with V2
3. In the options tab, select a supplier and choose the second currency
4. Save
5. In the options tab: select or unselect any supplier
6. See the symbol

``` 
Also this case

```php

1. Add new currency
2. add a product simple V2
3. in the options tab > select a supplier with a second supplier + price 1 DZD
4. Save
5. in the Pricing tab, the cost price = 1 euro
6. Create an order and check the table Prefix_order_detail

```
4. Another issue to create: when we delete all suppliers, the supplier block is still displayed => in the spec `This part is displayed only if you have created and enabled at least one supplier.`
5. Import combinations as commented here: https://github.com/PrestaShop/PrestaShop/pull/27353#pullrequestreview-963560433  **Second issue**
6. A cost price is empty => we have an alert, it is not automatically converted to 0.00 
![image](https://user-images.githubusercontent.com/16067358/167703345-7af4dc3f-af57-4d48-8850-3436ca1c5444.png)


Thanks @jolelievre, @MatShir and @HanaRebaiQA!

matks commented:
Thank you @jolelievre
